fix: revert shared library split on non-Windows platforms

Kha opened the pull request:
Avoids the performance hit and fixes #3528.

Kha commented:
#3500 was not sufficient

Kha commented:
@JLimperg Could you try `leanprover/lean4-pr-releases:pr-release-3529`?

Kha commented:
(ran out of time to manually test myself but `nm` output looks good so better to have this in the nightly than not)

JLimperg commented:
> @JLimperg Could you try `leanprover/lean4-pr-releases:pr-release-3529`?

The issue persists with this toolchain.

Kha commented:
@JLimperg Thanks for testing! I'm quite sure it's fixed now because I added a test and also reverted the change breaking it in the first place but feel free to try again!

JLimperg commented:
Can't reproduce any more. :party: Thank you very much for the fix!

Kha commented:
Thanks again for testing!
